go/mysqlconn/ldapauthserver/auth_server_ldap.go

```diff
@@ -0,0 +1,168 @@
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	log "github.com/golang/glog"
+	"github.com/youtube/vitess/go/mysqlconn"
+	"github.com/youtube/vitess/go/vt/servenv/grpcutils"
+	"gopkg.in/ldap.v2"
+)
+
+var (
+	ldapAuthConfigFile   = flag.String("mysql_ldap_auth_config_file", "", "JSON File from which to read LDAP server config.")
+	ldapAuthConfigString = flag.String("mysql_ldap_auth_config_string", "", "JSON representation of LDAP server config.")
+)
+
+// AuthServerLdap implements AuthServer with an LDAP backend
+// * include port in ldapServer, "ldap.example.com:386"
+type AuthServerLdap struct {
+	ldapServer    string
+	ldapCert      string
+	ldapKey       string
+	ldapCA        string
+	queryUser     string
+	queryPassword string
+	queryStr      string
+	getGroups     bool
+	groupQueryStr string
+}
+
+func init() {
+	if *ldapAuthConfigFile == "" && *ldapAuthConfigString == "" {
+		log.Infof("Not configuring AuthServerLdap because mysql_ldap_auth_config_file and mysql_ldap_auth_config_string are empty")
+		return
+	}
+	if *ldapAuthConfigFile != "" && *ldapAuthConfigString != "" {
+		log.Infof("Both mysql_ldap_auth_config_file and mysql_ldap_auth_config_string are non-empty, can only use one.")
+		return
+	}
+	ldapAuthServer := newAuthServerLdap()
+
+	data := []byte(*ldapAuthConfigString)
+	if *ldapAuthConfigFile != "" {
+		var err error
+		data, err = ioutil.ReadFile(*ldapAuthConfigFile)
+		if err != nil {
+			log.Fatalf("Failed to read mysql_ldap_auth_config_file: %v", err)
+		}
+	}
+	if err := json.Unmarshal(data, &ldapAuthServer); err != nil {
+		log.Fatalf("Error parsing AuthServerLdap config: %v", err)
+	}
+	mysqlconn.RegisterAuthServerImpl("ldap", ldapAuthServer)
+}
+
+func newAuthServerLdap() *AuthServerLdap {
+	return &AuthServerLdap{}
+}
+
+// UseClearText is always true for AuthServerLdap
+func (asl *AuthServerLdap) UseClearText() bool {
+	return true
+}
+
+// Salt is unimplemented for AuthServerLdap
+func (asl *AuthServerLdap) Salt() ([]byte, error) {
+	panic("unimplemented")
+}
+
+// ValidateHash is unimplemented for AuthServerLdap
+func (asl *AuthServerLdap) ValidateHash(salt []byte, user string, authResponse []byte) (string, error) {
+	panic("unimplemented")
+}
+
+// ValidateClearText connects to the LDAP server over TLS,
+// searches for the user, attempts to bind as that user with the supplied password,
+// and, if so configured, queries for the user's groups, returning them in a
+// comma-separated string.
+// In reality, it runs whatever queries are supplied. See data/test/mysql_ldap_auth_config.json for an example.
+// It is recommended that queryUser have read-only privileges on ldapServer
+func (asl *AuthServerLdap) ValidateClearText(username, password string) (string, error) {
+	conn, err := ldap.Dial("tcp", asl.ldapServer)
+	defer conn.Close()
+	if err != nil {
+		return "", err
+	}
+
+	// Reconnect with TLS
+	idx := strings.LastIndex(asl.ldapServer, ":") // allow users to (incorrectly) specify ipv6 without []
```

Comment:
Please use netutil.SplitHostPort here, it uses net.SplitHostPort and has a fallback, I think that's exactly what you are using.